File: 卡码/代码随想录/C++/二叉树/01-二叉树遍历方式/LeetCode_0144_0145_0094.cpp
//144-145-94.二叉树的前/后/中序遍历
#include <iostream>
#include <vector>
#include "../TreeNode.h"
using namespace std;

class Solution{
public:
	Solution(){}
	~Solution(){}
	//先序遍历
	vector<int> preorderTraversal(TreeNode* root) {
		vector<int> res;
		preOrder(root,res);
		return res;
	}
	void preOrder(TreeNode* root,vector<int>& v){
		if(root==nullptr) return;
		v.push_back(root->val);
		preOrder(root->left,v);
		preOrder(root->right,v);
	}
	//中序遍历
	vector<int> inorderTraversal(TreeNode* root) {
		vector<int> res;
		inOrder(root,res);
		return res;
	}
	void inOrder(TreeNode* root,vector<int>& v){
		if(root==nullptr) return;
		inOrder(root->left,v);
		v.push_back(root->val);
		inOrder(root->right,v);
	}
	//后序遍历
	vector<int> postorderTraversal(TreeNode* root) {
		vector<int> res;
		postOrder(root,res);
		return res;
	}
	void postOrder(TreeNode* root,vector<int>& v){
		if(root==nullptr) return;
		postOrder(root->left,v);
		postOrder(root->right,v);
		v.push_back(root->val);
	}
};

int main(){
	system("pause");
	return 0;
}

File: 卡码/代码随想录/Python/二叉树/01-二叉树遍历方式/LeetCode_0144_0145_0094.py
#144-145-94.二叉树的前/后/中序遍历
from typing import *
from ..TreeNode import *

class Solution:
	#先序遍历
  def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
		res:List[int]=[]
		self.preOrder(root,res)
		return res
	def preOrder(self,root:Optional[TreeNode],res:List[int]):
		if root==None:
			return
		res.append(root.val)
		self.preOrder(root.left,res)
		self.preOrder(root.right,res)
	#中序遍历
	def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
		res:List[int]=[]
		self.inOrder(root,res)
		return res
	def inOrder(self,root:Optional[TreeNode],res:List[int]):
		if root==None:
			return
		self.inOrder(root.left,res)
		res.append(root.val)
		self.inOrder(root.right,res)
	#后序遍历
	def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
		res:List[int]=[]
		self.postOrder(root,res)
		return res
	def postOrder(self,root:Optional[TreeNode],res:List[int]):
		if root==None:
			return
		self.postOrder(root.left,res)
		self.postOrder(root.right,res)
		res.append(root.val)
